### ISSUE TITLE
incorrect styles across component boundaries

### ISSUE THREAD

**Issue by jjenzz** [css-hooks/css-hooks#16]
i had a quick play with this and noticed that when there are conflicting longhand/shortand properties across component boundaries, it will [sometimes result in incorrect styles](https://codesandbox.io/p/devbox/charming-https-99txml?file=%2Fsrc%2FApp.tsx). i was expecting the `backgroundColor: black` to win here.

i haven't looked into the css-hooks code but on the surface it looks like it could be relying on object merging to override styles which can cause this issue:

![Screenshot 2023-12-01 at 00 24 01](https://github.com/css-hooks/css-hooks/assets/175330/822fcd6d-687d-4104-8bea-ff09df0c9682)

note how `background: 'orange'` comes **after** `backgroundColor: 'blue'` in the resulting object.

love what you're doing over here btw 👍 i have been exploring similar ideas with a css lib of my own and it's very reassuring to find others exploring `style` attribute solutions ☺️

**Comment by nsaunders** [comment]
Hi Jenna, thanks for raising this issue.

You are right to be concerned about mixing shorthand and longhand properties. CSS Hooks makes no effort to expand shorthand properties, which can lead to some surprising results. The project has gained popularity way more rapidly than anticipated, but I've had it in the back of my mind that there should probably be something like an ESLint plugin to help people avoid this pitfall.

With that said, if I'm understanding it correctly, the issue isn't caused by or limited to CSS Hooks. I've removed the `hooks()` calls in your example and am still seeing the unexpected result: [updated example](https://codesandbox.io/p/devbox/naughty-tristan-xck349?file=%2Fsrc%2FApp.tsx%3A12%2C1)

If I've missed something, please let me know!

**Comment by robinweser** [comment]
I'd say this is an issue not only with inline style but with CSS in general. I once wrote about it: https://weser.io/blog/the-shorthand-longhand-problem-in-atomic-css
There's https://github.com/robinweser/inline-style-expand-shorthand that helps with some of it, but I wouldn't recommend using it necessarily. Personally, I'd rather enforce longhands wining in whatever way. For CSS there're repeated selectors, for inline styles I assume we can just sort the object?

**Comment by jjenzz** [comment]
> the issue isn't caused by or limited to CSS Hooks

correct, this is a side-effect of `Object.assign` however, when people use styling libraries for composition across component boundaries the expectation is usually that "the last style applied always wins".
